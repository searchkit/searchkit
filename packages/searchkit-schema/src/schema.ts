import gql from 'graphql-tag'
import {
  FacetsResolver,
  HitsResolver,
  SearchkitConfig,
  SearchkitResolver,
  SummaryResolver
} from './resolvers'
import FacetResolver from './resolvers/FacetResolver'

const baseSearchkitTypeDefs = gql`
  type SKSortOption {
    id: ID!
    label: String!
  }

  type SKSummary {
    total: Float!
    appliedFilters: [SKSelectedFilter]
    query: String
    sortOptions: [SKSortOption]
  }

  interface SKSelectedFilter {
    id: ID!
    identifier: String!
    label: String!
    display: String!
  }

  type ValueSelectedFilter implements SKSelectedFilter {
    id: ID!
    identifier: String!
    display: String!
    label: String!
    value: String!
  }

  type NumericRangeSelectedFilter implements SKSelectedFilter {
    id: ID!
    identifier: String!
    label: String!
    display: String!
    min: Float!
    max: Float!
  }

  type DateRangeSelectedFilter implements SKSelectedFilter {
    id: ID!
    identifier: String!
    label: String!
    display: String!
    dateMin: String!
    dateMax: String!
  }

  type GeoBoundingBoxSelectedFilter implements SKSelectedFilter {
    id: ID!
    identifier: String!
    label: String!
    display: String!
    topLeft: SKGeoPoint
    bottomRight: SKGeoPoint
    bottomLeft: SKGeoPoint
    topRight: SKGeoPoint
  }

  type SKGeoPoint {
    lat: Float!
    lon: Float!
  }

  type SKPageInfo {
    total: Float
    totalPages: Float
    pageNumber: Float
    from: Float
    size: Float
  }

  type SKHitResults {
    items: [SKHit]
    page: SKPageInfo
    sortedBy: String
  }

  interface SKHit {
    id: ID!
  }

  input SKPageInput {
    from: Float
    size: Float
  }

  input SKFiltersSet {
    identifier: String!
    value: String
    min: Float
    max: Float
    dateMin: String
    dateMax: String
    geoBoundingBox: SKGeoBoundingBoxInput
  }

  input SKGeoBoundingBoxInput {
    topLeft: SKGeoPointInput
    bottomRight: SKGeoPointInput
    topRight: SKGeoPointInput
    bottomLeft: SKGeoPointInput
  }

  input SKGeoPointInput {
    lat: Float!
    lon: Float!
  }

  interface SKFacetSet {
    identifier: String
    label: String
    type: String
    display: String
    entries: [SKFacetSetEntry]
  }

  type SKFacetSetEntry {
    id: ID!
    label: String
    count: Float
    isSelected: Boolean
  }

  type RefinementSelectFacet implements SKFacetSet {
    identifier: String
    label: String
    type: String
    display: String
    entries: [SKFacetSetEntry]
  }

  type RangeFacet implements SKFacetSet {
    identifier: String
    label: String
    type: String
    display: String
    entries: [SKFacetSetEntry]
  }

  type DateRangeFacet implements SKFacetSet {
    identifier: String
    label: String
    type: String
    display: String
    entries: [SKFacetSetEntry]
  }
`

export interface SearchkitSchemaConfig {
  typeName: string
  hitTypeName: string
  config: SearchkitConfig
  addToQueryType?: boolean
  getBaseFilters?: any
}

export default (schemaConfigs: SearchkitSchemaConfig | Array<SearchkitSchemaConfig>) => {
  const { typeDefs, context, resolvers } = (Array.isArray(schemaConfigs)
    ? schemaConfigs
    : [schemaConfigs]
  ).reduce(
    (sum, schemaConfig) => {
      const typeName = schemaConfig.typeName

      const extendQuery = schemaConfig.addToQueryType
        ? `
    extend type Query {
      results(query: String, filters: [SKFiltersSet], page: SKPageInput): ${typeName}
    }`
        : ''

      const configTypeDefs = gql`
      type ${typeName} {
        summary: SKSummary
        hits(page: SKPageInput, sortBy: String): SKHitResults
        facets: [SKFacetSet]
        facet(identifier: String!, query: String, size: Float): SKFacetSet
      }

      ${extendQuery}
    `

      return {
        typeDefs: [...sum.typeDefs, configTypeDefs],
        resolvers: {
          ...sum.resolvers,
          [typeName]: {
            hits: HitsResolver,
            facets: FacetsResolver,
            facet: FacetResolver,
            summary: SummaryResolver
          },
          ...(schemaConfig.addToQueryType
            ? {
                Query: {
                  ...sum.resolvers.Query,
                  results: SearchkitResolver
                }
              }
            : {
                Query: sum.resolvers.Query
              })
        },
        context: {
          searchkit: {
            configs: {
              ...sum.context.searchkit.configs,
              [typeName]: schemaConfig.config
            },
            baseFilters: {
              ...sum.context.searchkit.baseFilters,
              [typeName]: schemaConfig.getBaseFilters
            },
            hitTypeMappings: {
              ...sum.context.searchkit.hitTypeMappings,
              [typeName]: schemaConfig.hitTypeName
            }
          }
        }
      }
    },
    {
      typeDefs: [baseSearchkitTypeDefs],
      resolvers: {
        SKHit: {
          __resolveType: (e) => e.type
        },
        SKFacetSet: {
          __resolveType: (e) => e.type
        },
        SKSelectedFilter: {
          __resolveType: (e) => e.type
        },
        Query: {}
      },
      context: {
        searchkit: {
          configs: {},
          baseFilters: {},
          hitTypeMappings: {}
        }
      }
    }
  )

  return {
    typeDefs,
    withSearchkitResolvers: (userResolvers: any = {}) => ({
      ...userResolvers,
      ...resolvers,
      Query: {
        ...userResolvers.Query,
        ...resolvers.Query
      }
    }),
    context
  }
}
