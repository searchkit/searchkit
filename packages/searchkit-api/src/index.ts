import type { MultipleQueriesQuery as AlgoliaMultipleQueriesQuery } from '@algolia/client-search'
import { transformRequest } from './transformRequest'
import transformResponse, { transformFacetValuesResponse } from './transformResponse'
import { ClientConfig, SearchRequest, RequestOptions, Transporter, AppSettings } from './types'
import { ESTransporter } from './Transporter'
import { getQueryRulesActionsFromRequest, QueryRuleActions } from './queryRules'
import { createElasticsearchQueryFromRequest } from './utils'
export * from './types'

class Client {
  transporter: Transporter

  constructor(private config: ClientConfig, private settings: AppSettings = { debug: false }) {
    this.transporter =
      'msearch' in config.connection ? config.connection : new ESTransporter(config.connection)
  }

  private async performSearch(requests: SearchRequest[]) {
    if (this.settings.debug) {
      console.log('Performing search with requests:')
      console.log('POST /_msearch')
      console.log(createElasticsearchQueryFromRequest(requests))
    }
    const responses = await this.transporter.msearch(requests)
    return responses
  }

  async handleRequest(body: any, requestOptions?: RequestOptions) {
    const instantsearchRequests = body as AlgoliaMultipleQueriesQuery[]

    const instantsearchResponses = this.handleInstantSearchRequests(
      instantsearchRequests,
      requestOptions
    )

    return instantsearchResponses
  }

  async handleInstantSearchRequests(
    instantsearchRequests: AlgoliaMultipleQueriesQuery[],
    requestOptions?: RequestOptions
  ) {
    const queryRules = this.config.search_settings.query_rules || []

    const requestQueryRuleActions: QueryRuleActions[] = instantsearchRequests.map((request) => {
      return getQueryRulesActionsFromRequest(queryRules, request)
    })

    let esRequests: SearchRequest[] = instantsearchRequests.map((request, i) => ({
      body: transformRequest(
        request,
        this.config.search_settings,
        requestQueryRuleActions[i],
        requestOptions
      ),
      request: request,
      indexName: request.indexName
    }))

    if (requestOptions?.hooks?.beforeSearch) {
      esRequests = await requestOptions.hooks.beforeSearch(esRequests)
    }

    let esResponses = await this.performSearch(esRequests)

    if (requestOptions?.hooks?.afterSearch) {
      esResponses = await requestOptions.hooks.afterSearch(esRequests, esResponses)
    }

    const instantsearchResponses = esResponses.map((response, i) => {
      // @ts-ignore
      if (instantsearchRequests[i].params?.facetName) {
        return transformFacetValuesResponse(response, instantsearchRequests[i])
      }
      return transformResponse(
        response,
        instantsearchRequests[i],
        this.config.search_settings,
        requestQueryRuleActions[i]
      )
    })

    return {
      results: instantsearchResponses
    }
  }
}

const createClient = (config: ClientConfig, settings?: AppSettings) => new Client(config, settings)

export default createClient
