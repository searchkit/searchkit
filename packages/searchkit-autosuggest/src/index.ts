export * from "./SearchkitAutosuggest"
export * from "./datasources"
export * from "./Utils"