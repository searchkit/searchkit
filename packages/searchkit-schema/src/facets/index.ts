export { default as RefinementSelectFacet } from './RefinementSelectFacet'
export { default as RangeFacet } from './RangeFacet'
export { default as DateRangeFacet } from './DateRangeFacet'
export { default as HierarchicalMenuFacet } from './HierarchicalMenuFacet'
export * from './BaseFacet'
export * from './VisibilityRules'
