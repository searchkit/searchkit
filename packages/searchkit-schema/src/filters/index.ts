export { default as GeoBoundingBoxFilter } from './GeoBoundingBoxFilter'
export { default as TermFilter } from './TermFilter'
export { default as NumericRangeFilter } from './NumericRangeFilter'
export * from './BaseFilter'
