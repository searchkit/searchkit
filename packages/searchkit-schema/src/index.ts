export * from './resolvers'
export * from './query'
export * from './facets'
export * from './filters'

export { default as SearchkitSchema, SearchkitSchemaConfig } from './schema'
