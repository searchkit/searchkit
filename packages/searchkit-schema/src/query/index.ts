export { default as MultiMatchQuery } from './MultiMatchQuery'
export { default as BaseQuery } from './BaseQuery'
export { default as CustomQuery } from './CustomQuery'
