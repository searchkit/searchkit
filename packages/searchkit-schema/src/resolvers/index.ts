export { default as HitsResolver } from './HitsResolver'
export { default as SearchkitResolver } from './ResultsResolver'
export { default as FacetsResolver } from './FacetsResolver'
export { default as SummaryResolver } from './SummaryResolver'
export { default as FacetResolver } from './FacetResolver'
