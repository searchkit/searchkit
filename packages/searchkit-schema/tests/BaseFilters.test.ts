import nock from 'nock'
import { SearchkitConfig, MultiMatchQuery } from '../src/index'
import { setupTestServer, callQuery } from './support/helper'
import HitsMock from './__mock-data__/HitResolver/Hits.json'

describe('Base Query', () => {
  const runQuery = async (query = '', page = { size: 10, from: 0 }, sorting?: string) => {
    const gql = `
        {
          results( query: "heat", filters: [{ identifier: "type", value: "movie"}]) {
            hits {
              items {
                ... on ResultHit {
                  id
                  fields {
                    writers
                    actors
                  }
                }
              }
            }
          }
        }
      `
    const response = await callQuery({ gql })
    return response
  }

  it('should return correct Results', async () => {
    const moviesSearchConfig: SearchkitConfig = {
      host: 'http://localhost:9200',
      index: 'movies',
      hits: {
        fields: ['actors', 'writers']
      },
      query: new MultiMatchQuery({ fields: ['actors', 'writers', 'title^4', 'plot'] })
    }

    setupTestServer({
      typeName: 'ResultSet',
      hitTypeName: 'ResultHit',
      config: moviesSearchConfig,
      addToQueryType: true,
      getBaseFilters: (parent, parameters, ctx) => {
        expect(parent).toBe(undefined)
        expect(parameters).toBeDefined()
        expect(ctx).toBeDefined()
        return [{ term: { status: 'published' } }]
      }
    })

    const scope = nock('http://localhost:9200')
      .post('/movies/_search')
      .reply((uri, body) => {
        expect(body).toMatchInlineSnapshot(`
          Object {
            "_source": Object {
              "includes": Array [
                "actors",
                "writers",
              ],
            },
            "aggs": Object {},
            "from": 0,
            "query": Object {
              "bool": Object {
                "filter": Array [
                  Object {
                    "term": Object {
                      "status": "published",
                    },
                  },
                ],
                "should": Array [
                  Object {
                    "multi_match": Object {
                      "fields": Array [
                        "actors",
                        "writers",
                        "title^4",
                        "plot",
                      ],
                      "operator": "and",
                      "query": "heat",
                      "type": "best_fields",
                    },
                  },
                  Object {
                    "multi_match": Object {
                      "fields": Array [
                        "actors",
                        "writers",
                        "title^4",
                        "plot",
                      ],
                      "query": "heat",
                      "type": "cross_fields",
                    },
                  },
                  Object {
                    "multi_match": Object {
                      "fields": Array [
                        "actors",
                        "writers",
                        "title^4",
                        "plot",
                      ],
                      "query": "heat",
                      "type": "phrase",
                    },
                  },
                  Object {
                    "multi_match": Object {
                      "fields": Array [
                        "actors",
                        "writers",
                        "title^4",
                        "plot",
                      ],
                      "query": "heat",
                      "type": "phrase_prefix",
                    },
                  },
                ],
              },
            },
            "size": 10,
            "sort": Array [
              Object {
                "_score": "desc",
              },
            ],
          }
        `)
        return [200, HitsMock]
      })

    const response = await runQuery()
    expect(response.body.data).toMatchSnapshot('userhits')
    expect(response.status).toEqual(200)
  })
})
