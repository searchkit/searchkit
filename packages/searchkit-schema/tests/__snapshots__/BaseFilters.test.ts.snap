// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Base Query should return correct Results: userhits 1`] = `
Object {
  "results": Object {
    "hits": Object {
      "items": Array [
        Object {
          "fields": Object {
            "actors": Array [
              "Christian Bale",
              "Heath Ledger",
              "Aaron Eckhart",
              "Michael Caine",
            ],
            "writers": Array [
              "Jonathan Nolan",
              "Christopher Nolan",
              "David S. Goyer",
              "Bob Kane",
            ],
          },
          "id": "tt0468569",
        },
        Object {
          "fields": Object {
            "actors": Array [
              "Tim Roth",
              "Amanda Plummer",
              "Laura Lovelace",
              "John Travolta",
            ],
            "writers": Array [
              "Quentin Tarantino",
              "Roger Avary",
            ],
          },
          "id": "tt0110912",
        },
        Object {
          "fields": Object {
            "actors": Array [
              "Noel Appleby",
              "Alexandra Astin",
              "Sean Astin",
              "David Aston",
            ],
            "writers": Array [
              "J.R.R. Tolkien",
              "Fran Walsh",
              "Philippa Boyens",
              "Peter Jackson",
            ],
          },
          "id": "tt0167260",
        },
        Object {
          "fields": Object {
            "actors": Array [
              "Morgan Freeman",
              "Andrew Kevin Walker",
              "Daniel Zacapa",
              "Brad Pitt",
            ],
            "writers": Array [
              "Andrew Kevin Walker",
            ],
          },
          "id": "tt0114369",
        },
        Object {
          "fields": Object {
            "actors": Array [
              "Michael Berryman",
              "Peter Brocco",
              "Dean R. Brooks",
              "Alonzo Brown",
            ],
            "writers": Array [
              "Lawrence Hauben",
              "Bo Goldman",
              "Ken Kesey",
              "Dale Wasserman",
            ],
          },
          "id": "tt0073486",
        },
        Object {
          "fields": Object {
            "actors": Array [
              "Russell Crowe",
              "Joaquin Phoenix",
              "Connie Nielsen",
              "Oliver Reed",
            ],
            "writers": Array [
              "David Franzoni",
              "John Logan",
              "William Nicholson",
            ],
          },
          "id": "tt0172495",
        },
        Object {
          "fields": Object {
            "actors": Array [
              "Andrew Lincoln",
              "Steven Yeun",
              "Chandler Riggs",
              "Norman Reedus",
            ],
            "writers": Array [
              "Frank Darabont",
            ],
          },
          "id": "tt1520211",
        },
        Object {
          "fields": Object {
            "actors": Array [
              "Jim Carrey",
              "Kate Winslet",
              "Gerry Robert Byrne",
              "Elijah Wood",
            ],
            "writers": Array [
              "Charlie Kaufman",
              "Michel Gondry",
              "Pierre Bismuth",
            ],
          },
          "id": "tt0338013",
        },
        Object {
          "fields": Object {
            "actors": Array [
              "Marlon Brando",
              "Martin Sheen",
              "Robert Duvall",
              "Frederic Forrest",
            ],
            "writers": Array [
              "John Milius",
              "Francis Ford Coppola",
              "Michael Herr",
            ],
          },
          "id": "tt0078788",
        },
        Object {
          "fields": Object {
            "actors": Array [
              "Ellen Burstyn",
              "Jared Leto",
              "Jennifer Connelly",
              "Marlon Wayans",
            ],
            "writers": Array [
              "Hubert Selby Jr.",
              "Darren Aronofsky",
            ],
          },
          "id": "tt0180093",
        },
      ],
    },
  },
}
`;
